api/v1alpha1/trafficmanager_types.go

```diff
@@ -0,0 +1,264 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type TrafficManagerProfile struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec TrafficManagerProfileSpec `json:"spec"`
+	// +optional
+	Status TrafficManagerProfileStatus `json:"status,omitempty"`
+}
+
+type TrafficManagerProfileSpec struct {
+	// Type of routing method.
+	// +kubebuilder:validation:Enum=Weighted
+	// +kubebuilder:default=Weighted
+	// +optional
+	RoutingMethod TrafficRoutingMethod
+
+	// The DNS settings of the Traffic Manager profile.
+	// +required
+	DNSConfig *DNSConfig
+
+	// The endpoint monitoring settings of the Traffic Manager profile.
+	MonitorConfig *MonitorConfig
+
+	// The list of endpoints in the Traffic Manager profile.
+	Endpoints []*TrafficManagerEndpointRef
+}
+
+type TrafficManagerEndpointRefType string
+
+const (
+	// TrafficManagerEndpointTypeServiceImport is a group of public ip exposed by the exported service.
+	TrafficManagerEndpointTypeServiceImport TrafficManagerEndpointRefType = "ServiceImport"
+)
+
+type TrafficManagerEndpointRef struct {
+	// Type of endpoint reference. Can be "ServiceImport". Default is ServiceImport.
+	// +kubebuilder:validation:Enum=ServiceImport
+	// +kubebuilder:default=ServiceImport
+	// +optional
+	Type TrafficManagerEndpointRefType
+
+	ServiceImportEndpointConfig *ServiceImportEndpointConfig
+}
+
+type TrafficManagerEndpointType string
+
+const (
+	// EndpointTypeAzureEndpoints are used for services hosted in Azure exposed via PublicIPAddress.
+	// The publicIpAddress must have a DNS name assigned to be used in a Traffic Manager profile.
+	EndpointTypeAzureEndpoints TrafficManagerEndpointType = "AzureEndpoints"
+	// EndpointTypeExternalEndpoints are used for IPv4/IPv6 addresses, FQDNs, or for services hosted outside Azure.
+	// Theses services can either be on-premises or with a different hosting provider.
+	EndpointTypeExternalEndpoints TrafficManagerEndpointType = "ExternalEndpoints"
+)
+
+type ServiceImportEndpointConfig struct {
+	// ServiceImport is the reference to the ServiceImport in the same namespace of Traffic Manager Profile.
+	// +required
+	ServiceImport *ServiceImportRef
+
+	// +required
+	Config *TrafficManagerEndpoint
+}
+
+type TrafficManagerEndpoint struct {
+	// If the type is "AzureEndpoints", it means the public ip is an azure resource and must have a DNS name assigned.
+	// Note: To use Traffic Manager with endpoints from other subscriptions, the controller needs to have read access to
+	// the endpoint.
+	// +kubebuilder:validation:Enum=AzureEndpoints;ExternalEndpoints
+	// +kubebuilder:default=AzureEndpoints
+	// +optional
+	Type TrafficManagerEndpointType
+
+	// If Always Serve is enabled, probing for endpoint health will be disabled and endpoints will be included in the traffic
+	// routing method.
+	// +kubebuilder:default=false
+	// +optional
+	AlwaysServe bool
```

Comment:
I assume it's a direct translation of an ATM field?

Reply:
yes, required by atm to support weighted routing method